python/paddle/distributed/auto_parallel/api.py

```diff
@@ -1566,3 +1601,307 @@ def unshard_dtensor(dist_tensor):
         default_dist_ctx._dist_tensors_for_program.pop(serial_tensor_id, None)
 
         return dist_tensor
+
+
+class ShardDataloader:
+    """
+    ShardDataloader converts a dataloader to a new dataloader which provided two capabilities:
+    1. split dataloader by shard_dim to do data parallel.
+    2. reshard the output of dataloader to distributed tensor.
+    if is_dataset_splitted is True, just need to do reshard.
+
+    Args:
+        dataloader (paddle.io.DataLoader): The dataloader to be sharded.
+        meshes (ProcessMesh|list[ProcessMesh]|tuple[ProcessMesh]): The mesh list of the dataloader.
+            Identify which mesh the input is on. if len(meshes) == 1 or type(meshes) == ProcessMesh,
+            all the inputs are on the same mesh.
+        input_keys (list[str]|tuple[str]): if the iteration result of dataloader is a dict of tensors,
+            input_keys is the keys of this dict, identify which tensor is located on which mesh,
+            one-to-one correspondence with meshes. i.e. dict[input_keys[i]] is on meshes[i].
+            Default: None, which means the outputs is a list, and the i'th input is on meshes[i].
+        shard_dims (list|tuple|str|int]): The mesh dimension to shard the dataloader.
+            Users can specify the shard_dim of each mesh or specify a single shard_dim for all meshes.
+            Default: None, which means the data loader will not be split, i.e. mp.
+        is_dataset_splitted (bool): Whether the dataset has been splitted.
+    """
+
+    def __init__(
+        self,
+        dataloader: paddle.io.DataLoader,
+        meshes: Union[
+            dist.ProcessMesh, List[dist.ProcessMesh], Tuple[dist.ProcessMesh]
+        ],
+        input_keys: Union[List[str], Tuple[str]] = None,
+        shard_dims: Union[list, tuple, str, int] = None,
+        is_dataset_splitted: bool = False,
+    ):
+        # do some check
+        if is_dataset_splitted is True and shard_dims is None:
+            raise ValueError(
+                "shard_dims must be set when is_dataset_splitted is True"
+            )
+
+        self._meshes = to_list(meshes)
+        if self._meshes is None or len(self._meshes) == 0:
+            raise ValueError("meshes must be set")
+
+        process_id = dist.get_rank()
+        if self._process_id_in_multi_meshes(process_id):
+            raise ValueError("process_id is in more than one mesh")
```

Comment:
"process_id %d is in more than one mesh, the meshes are %s" % process_id, meshes

Reply:
Done